Add 1-Dimensional memory model for memory safe tasks

jankoerner opened the pull request:
This pull request introduces a 1-dimensional memory model which can be used for ReachSafety tasks.
The 1-dimensional memory model comes with some major refactorings of the source code.
These refactoring can be seen in the attached image and allow an easy extension for new Addressings, Structures, ManagementStrategies and Metadatas (not visible in the image). 
<img width="1904" height="905" alt="2025-11-06T20:16:22,830512837+01:00" src="https://github.com/user-attachments/assets/7baa40ef-3b8d-42f2-ab08-6d6ed1865f03" />

jankoerner committed:
CACSL2Boogie: Renamed MemoryModel

Renamed memory model to memory structure because it only covers the
memory representation. The full memory model alos features a memory
addressing and this commit is a preparation for this step.

jankoerner committed:
CACSL2Boogie: Added memory addressing settings

jankoerner committed:
CACSL2Boogie: Added files for memory model refactor

Added the interfaces for structure and addressing.
Added 1D and 2D memory addressing as well as a common base class.
Added a utils class.

jankoerner committed:
CACSL2Boogie: Removed ExpressionTranslation

The ExpressionTranslation was neither used in the abstract base class nor in the subclasses.
Also removed unnecessary cast of TypeHandler.

jankoerner committed:
CACSL2Boogie: Added MemoryModelFactory

This factory is used to create concrete instances implementing
interfaces that are used in the memory model. All validity checks are
also part of the factory.

jankoerner committed:
CACSL2Boogie: Use memory model in memory handler

Use the memory structure from the new memory model instead of the using
the structure that was part of the memory handler.
From time to time, more parts will move to the memory model.

jankoerner committed:
CACSL2Boogie: Use IMemoryStructure

jankoerner committed:
CACSL2Boogie: Moved malloc specifications to model

The malloc now depent on the settings.
Moved things out of the MemoryHandler to access them there and in the
model.
Renamed things.

jankoerner committed:
CACSL2Boogie: Moved dealloc specifications to model

jankoerner committed:
CACSL2Boogie: Moved init statements to model

The statements in Ultimate.Init depent on the current memory model.

jankoerner committed:
CACSL2Boogie: Moved allocInit specs to model

jankoerner committed:
CACSL2Boogie: Moved BooleanArrayHelper to own file

Moved the IBooleanArrayHelper and the interfaces to seperate files,
since they are also used outside of the MemoryHandler.

jankoerner committed:
CACSL2Boogie: Removed unused variables

jankoerner committed:
CACSL2Boogie: Moved MemoryArea to own file

It is also used outsided of the MemoryHandler.

jankoerner committed:
CACSL2Boogie: Setting for FixedAddressForInit

This option can be activated or deactivate through a setting instead of
a constant in the MemoryHandler.

jankoerner committed:
CACSL2Boogie: Moved pointerArithmetic to mem model

Also removed manual pointer arithmetic in the InitializationHandler and
called the corresponding doPointerArithmetic() function of the
MemoryHandler instead.

jankoerner committed:
CACSL2Boogie: AllocInit Counter to memory model

The the step size of the base address counter incremention depends now
on the exact implementation of the memory model.
With the 1D-addressing mode, the step size always increase with the
total size of allocated object.
With the 2D-addressing mode, it increases with a step size of one per
allocated object.

jankoerner committed:
CACSL2Boogie: Added assume statement to init

If using the 1D-memory addressing the assume statement for the
\#InitialAllocation counter is added to Ultimate.init.

jankoerner committed:
CACSL2Boogie: Fixed warning

jankoerner committed:
CACSL2Boogie: Moved structFieldAddress to Model

The field addresses of structs are now calculated in the memory model.

jankoerner committed:
CACSL2Boogie: Settings for regression tests

To test the 1D model, it is set to default.
For the Memsafety regression tests, the 2D model is still used.

jankoerner committed:
CACSL2Boogie: Moved addIntegerToPointer to model

jankoerner committed:
CACSL2Boogie: Removed unnecessary if

This if clause is already checked in the getPointerBaseAddress and the
getPointerOffset methods.

jankoerner committed:
CACSL2Boogie: Moved the ptr base validity check

jankoerner committed:
CACSL2Boogie: Simplified expression creations

jankoerner committed:
CACSL2Boogie: Move fully allocated check to model

jankoerner committed:
CACSL2Boogie: Moved freed pointer checks to model

jankoerner committed:
CACSL2Boogie: Fixed getWriteCallStruct

This method did the address calculation for a struct field manually,
which is now done by the memory model.

jankoerner committed:
CACSL2Boogie: 1D Model compatible with bitvector

jankoerner committed:
CACSL2Boogie: Workaround for Regression Tests

Currently many regression tests fail, because MemSafety checks are used,
which are not allowed within the 1D memory addressing mode.
If any MemSafety check is active, instead of an exception the 2D model
is returned.

jankoerner committed:
CACSL2Boogie: Added incompatible 1D setting

Within the 1D memory addressing we are not able to check the allocation
purity because we don't have the #valid array.

jankoerner committed:
CACSL2Boogie: Changed default translation settings

Changed the default translation settings, because the 1D memory
addressing is not compatible with any MemSafety checks, which have been
enabled as default.

jankoerner committed:
CACSL2Boogie: Fixed getWriteCallArray

The address calculation was done manually, is now done in the memory
model using the addIntegerConstantToPointer in the memory model.
Do the same in the ExpressionResultTransformer as well, where the
data is manually read from a heap array.

jankoerner committed:
CACSL2Boogie: Fixed readStructFromHeap

The address was calculated manually. Uses now the function in the memory
model for this.

jankoerner committed:
CACSL2Boogie: Fix Bug in 1D addressing

The StackAllocCounter was always returned for both STACK and HEAP
allocations.

jankoerner committed:
CACSL2Boogie: Use var instead of function call

jankoerner committed:
CACSL2Boogie: Updated comments to match 2D

jankoerner committed:
CACSL2Boogie: Fixed TypeCheckException in 1D

Fixed the TypeCheckException in the 1D memory addressing mode, which
occured during the nightly regression tests e.g. in: I_programs_regression_c_nestedStruct-Safe.c
S_programs_regression_c_AutomizerC-BitvectorTranslation.epf T_programs_regression_c_AutomizerC.xml
– de.uni_freiburg.informatik.ultimate.regressiontest.generic.RegressionTestSuite

jankoerner committed:
CACSL2Boogie: Make methods static

jankoerner committed:
CACSL2Boogie: Added comments

jankoerner committed:
CACSL2Boogie: Introduce variable for size_t

jankoerner committed:
CACSL2Boogie: Introduced Pointer Type

jankoerner committed:
CACSL2Boogie: Moved varargs to the model

jankoerner committed:
CACSL2Boogie: Removed unsused member

jankoerner committed:
CACSL2Boogie: Moved lastCharOfString to model

jankoerner committed:
CACSL2Boogie: Replaced manual pointer arithmetic

jankoerner committed:
CACSL2Boogie: Created initialPointerFromPointer

jankoerner committed:
CASCL2Boogie: Moved strchr assume to the model

jankoerner committed:
CACSL2Boogie: Constant arrays are incomp. options

The constant array optimization cannot be used using the 1D memory
model.

jankoerner committed:
CACSL2Boogie: Used create null pointer method

jankoerner committed:
CACSL2Boogie: Removed empty comment

jankoerner committed:
CACSL2Boogie: Moved memsafety checks for pointers

Moved the memsafety checks to the memory model

jankoerner committed:
CACSL2Boogie: Removed unused parameter

jankoerner committed:
CACSL2Boogie: Fixed warnings

Supress some warning sometimes because some things my be used again.

jankoerner committed:
CACSL2Boogie: Moved overlapping string checks

jankoerner committed:
CACSL2Boogie: Moved rhs of array assignment

jankoerner committed:
CACSL2Boogie: 1D Adressing implemented

Introduced a memory pointer interface that is part of the typehandler.
Moved all pointer creation out of the memory model to this interface, or
the classes implementing this interface.

jankoerner committed:
CACSL2Boogie: Hide memory structure in model

jankoerner committed:
CACSL2Boogie: Moved function to specialization

This function was only used in one derived class, so it was moved there.

jankoerner committed:
CACSL2Boogie: Removed abstract declaration

This function is already part of the interface, so it is unnecessary to
declare it as abstract again.

jankoerner committed:
CACSL2Boogie: Added factory for memory model

Added a factory method that creates a valid memory model.
Additionally added an enum representing all valid combinations of memory
structure and addressing.

jankoerner committed:
CACSL2Boogie: Moved default implementation to base

Moved the default implementations of methods that are not supported by
all addressing modes to the base class. From now on, new modes don't
need to override these functions if they dont support them.

jankoerner committed:
CACSL2Boogie: Removed base of the interfaces

Renamed the interface functions s.t. they don't use the term base
anymore.

jankoerner committed:
CACSL2Boogie: Added IMemoryManagementStrategy

Extracted the memory management strategy, e.g. how memory is allocated
and deallocated from the addressing in order to allow the implementation
of different strategies (e.g. 1D addressing, count up/down).

jankoerner committed:
CACSL2Boogie: Restored default settings

Restored the default Ultimate settings, which have been used before
different addressing modes.
Additionally removed the workaround to fix the regression tests, since
it is not needed anymore.

jankoerner committed:
CACSL2Boogie: Removed comment

jankoerner committed:
CACSL2Boogie: Added 1D Translation settings

Added settings files for the usage of the 1D memory model in the nightly
regression tests of CToBoogieTranlation test suite, because the default
setting was changed to 2D.

jankoerner committed:
CACSL2Boogie: Added 1D Reach regression test

Added a settings file for reachsafety regression tests with 1D memory
addressing.
Used the settings file for svcomp as a starting point.

jankoerner committed:
CACSL2Boogie: Fixed bug with varargs end

In the boogie translation for the varargs, the call to Ultimate.dealloc
was not correct, since it was called with "ULTIMATE.dealloc({ base: ~v~0 });"
instead of "ULTIMATE.dealloc({ base: ~v~0!base });" leading to a type
error in the Preprocessor.

jankoerner committed:
CACSL2Boogie: Expanded .testignore

Added the regression test cases that fail b.c. of overapproximation for the 1D tests to the .testignore.

jankoerner committed:
CACSL2Boogie: Moved .testignore case

The ignore case was in the wrong .testignore file.

jankoerner committed:
CACSL2Boogie: Fixed translation error

Fixed a translation error, because within the 1D-model a offset value
was tried to access.

jankoerner committed:
CACSL2Boogie: Moved pointer subtraction to model

Moved the pointer subtraction into the memory model.
For 1D the base addresses are subtracted, the implementation for 2D
remains the same and offsets are subtracted.

jankoerner committed:
CACSL2Boogie: Removed explicit offset access

Removed the explicit access of the offset field in a struct, to hide in
the memory pointer.

jankoerner committed:
CACSL2Boogie: Added nullpointer check to interface

Added the nullpointer check to the IMemoryPointer interface, since it
differs for every implementation.

jankoerner committed:
CACSL2Boogie: Renamed variables

jankoerner committed:
CACSL2Boogie: Moved realloc statements to model

The realloc implementation of the 2D memory addressing is not applicable
for the 1D addressing. The 1D addressing needs to be implemented.

jankoerner committed:
CACSL2Boogie: Moved create to MemoryModel

Moved the factory method that creates the MemoryModel to the class
itself and hide the constructor. Now it is enforced that only valid
combinations of the memory model can be created.

jankoerner committed:
CACSL2Boogie: Hide memory pointer constructor

Hide the memory pointer constructor with a static create method in each
class in order to ensure that the 1D pointer is only created when the
settings are compatible.

jankoerner committed:
CACSL2Boogie: Added comments to realloc

jankoerner committed:
CASCL2Boogie: Added 1D realloc

jankoerner committed:
CACSL2Boogie: Renamed variables

jankoerner committed:
CACSL2Boogie: Fix 1D realloc

There were duplicate calls to memcpy, for each heapdata-array one, which
is not necessary, because memcpy already handles this.

jankoerner committed:
CACSL2Boogie: Extended alloc ensures

Extended the ensure clauses of allocOnHeap and allocOnStack. There was a
problem, causing the heap getting greater than the StackHeapBarrier,
leading to incorrect verification results.

jankoerner committed:
CACSL2Boogie: Rename setting Memory structure

Rename the setting from Memory Structure to Memory structure.

jankoerner committed:
CACSL2Boogie: Fixed 1D pointer relation

One dimensional pointers use their base address only to express a
relation. So the base equality is not needed in this case.
Fixes /c/loops/verisec_NetBSD-libc_loop.yml of the sv-comp.

jankoerner committed:
CACSL2Boogie: Fixed doPointerArithmetic

It did not exactly behave like before my changes but now it should.

jankoerner committed:
CACSL2Boogie: Added free to dealloc ensure

Added the free keyword to the dealloc ensure clause of the non-det
memory management strategy.

jankoerner committed:
CACSL2Boogie: Fixed structs in arrays

There was problem that led to a Classcast exception when the program
contained an array full of structs.

jankoerner committed:
CACSL2Boogie: 1D Added assume + ensures clause

Added an assume in Ultimate.init, that states that the Stackheapbarrier
is greater than the InitialAllocationsCounter.

jankoerner committed:
CACSL2Boogie: Fixed pthreads null-pointer check

Within the pthread library a there was check made to see if a pointer is
a null pointer, that only supported the 2D model.

jankoerner committed:
CACSL2Boogie: Fixed bug in bitvector translations

The counter and addresses during the bitvector translation have 32bit,
or 64bit values.
There are some incorrect results coming from the fact, that the chosen
addresses on the stack exceed the limits of 32-bit or 64-bit numerical
values.

jankoerner committed:
CACSL2Boogie: Backtranslator for 1D addressing

jankoerner committed:
CACSL2Boogie: Renamed classes

Renamed from OneDimensional... to ...1D and ...2D.

jankoerner committed:
CACSL2Boogie: Added IMemoryData interface

Added an interface that allows the creation of different metadata.
The memory management strategy depends on the metadata.
Also some memory safety checks are dependent on the metadata.

jankoerner committed:
Update settings files to match refactoring

Replaced the "memory model" with "memory structure" since the setting
was renamed during the implementation of the 1-dimensional addressing.

jankoerner committed:
Added witness generation validation tests 1D

Added witness generation and validation tests for the 1-dimensional
addressing.

jankoerner committed:
Added backtranslation regression tests for 1D

schuessf committed:
Fix compilation: Use memory neutrality setting

schuessf commented:
@jankoerner Thanks for your work! I will try to review this PR tomorrow. I just quickly fixed the compilation, as some settings were changed in #752 and started the Jenkins nightly for this PR.

bahnwaerter commented:
Many thanks also for your extensive work, @jankoerner!


> [...] I just quickly fixed the compilation, as some settings were changed in #752 and started the Jenkins nightly for this PR [...]

Thanks @schuessf for fixing the issue. It looked like the Git rebase didn't include all the other changes.

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/BooleanArrayHelper_Bitvector.java:
```diff
@@ -0,0 +1,37 @@
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.LocationFactory;
+import de.uni_freiburg.informatik.ultimate.core.model.models.ILocation;
+
+public final class BooleanArrayHelper_Bitvector implements IBooleanArrayHelper {
```
Underscores in class names should be avoided according to the Java camel case style. This also applies to the name of this source code file.

jankoerner replied:
See [731552f](https://github.com/ultimate-pa/ultimate/pull/756/commits/731552fefc90a1b0f26ca07476cb7eaa857938d4)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/BooleanArrayHelper_Bool.java:
```diff
@@ -0,0 +1,36 @@
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.LocationFactory;
+import de.uni_freiburg.informatik.ultimate.core.model.models.ILocation;
+
+public final class BooleanArrayHelper_Bool implements IBooleanArrayHelper {
```
Underscores in class names should be avoided according to the Java camel case style. This also applies to the name of this source code file.

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/BooleanArrayHelper_Integer.java:
```diff
@@ -0,0 +1,38 @@
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.LocationFactory;
+import de.uni_freiburg.informatik.ultimate.core.model.models.ILocation;
+
+public final class BooleanArrayHelper_Integer implements IBooleanArrayHelper {
```
Underscores in class names should be avoided according to the Java camel case style. This also applies to the name of this source code file.

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/CCharacterConstant.java:
```diff
@@ -104,6 +106,7 @@ public CCharacterConstant(final String sourceLiteral, final Signedness signednes
 		mRepresentingValue = computeRepresentingValue(mNumericalValue, signednessOfChar, mSignednessOfRepresentingType);
 	}
 
+	@SuppressWarnings("static-method")
```
Why not make the method static?

jankoerner replied:
[43a1ccc](https://github.com/ultimate-pa/ultimate/pull/756/commits/43a1ccc0922c891e1410139d9e054212c55d4d3d)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/IBooleanArrayHelper.java:
```diff
@@ -0,0 +1,18 @@
+
+	Expression compareWithTrue(Expression expr);
+
+	default Expression constructValue(final boolean value) {
```
Perhaps the interface method should be renamed to `constructFromValue`?

jankoerner replied:
See [bf2a925](https://github.com/ultimate-pa/ultimate/commit/bf2a9251bf23c01929004e0a22a407d1c43a4f33)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/IMemoryAdressing.java:
```diff
@@ -0,0 +1,251 @@
+	 *
+	 * @return The declarations.
+	 */
+	List<MemoryModelDeclarations> metaDataDeclarations();
```
Add prefix `get` to interface method name.

```suggestion
	List<MemoryModelDeclarations> getMetaDataDeclarations();
```

jankoerner replied:
See [bf2a925](https://github.com/ultimate-pa/ultimate/commit/bf2a9251bf23c01929004e0a22a407d1c43a4f33)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/IMemoryAdressing.java:
```diff
@@ -0,0 +1,251 @@
+	 *
+	 * @return The step size.
+	 */
+	BigInteger fixedAddressCounterCountingStep(final Expression size);
```
Add prefix `get` to interface method name.

```suggestion
	BigInteger getFixedAddressCounterCountingStep(final Expression size);
```

jankoerner replied:
See [bf2a925](https://github.com/ultimate-pa/ultimate/commit/bf2a9251bf23c01929004e0a22a407d1c43a4f33)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/IMemoryAdressing.java:
```diff
@@ -0,0 +1,251 @@
+	 *
+	 * @return The initial pointer.
+	 */
+	Expression initialPointerFromPointer(final ILocation loc, final Expression ptr);
```
Add prefix `construct` to interface method name.

```suggestion
	Expression constructInitialPointerFromPointer(final ILocation loc, final Expression ptr);
```

jankoerner replied:
See [bf2a925](https://github.com/ultimate-pa/ultimate/commit/bf2a9251bf23c01929004e0a22a407d1c43a4f33)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/IMemoryAdressing.java:
The names of the interface methods should be named consistently. Whenever an expression is generated, this should be clearly indicated in the name (for example with the prefix `construct`).

jankoerner replied:
See [bf2a925](https://github.com/ultimate-pa/ultimate/commit/bf2a9251bf23c01929004e0a22a407d1c43a4f33)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/IMemoryPointer.java:
The names of the interface methods should be named consistently. Whenever an expression is generated, this should be clearly indicated in the name (for example with the prefix construct).

jankoerner replied:
See [bf2a925](https://github.com/ultimate-pa/ultimate/commit/bf2a9251bf23c01929004e0a22a407d1c43a4f33)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryAddressing2D.java:
```diff
@@ -0,0 +1,529 @@
+				ExpressionFactory.newBinaryExpression(loc, Operator.LOGICAND, baseEqualsNull, offsetEqualsNull);
+		// old solution did not work quickly..
+		// final BinaryExpression equalsNull = expressionTranslation.constructBinaryComparisonExpression(loc,
+		// new BinaryExpression(loc, Operator.COMPEQ, tmpExpr, nullExpr);
```
Remove commented-out code!

```suggestion
```

jankoerner replied:
[a39f954](https://github.com/ultimate-pa/ultimate/pull/756/commits/a39f9548335512d1fbb9a9de2da878c6746d23d5)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryAdressingBase.java:
```diff
@@ -0,0 +1,224 @@
+	T mMemoryPointer;
+	IMemoryManagementStrategy mMemoryManagementStrategy;
+
+	BigInteger functionPointerPointerBaseValue = BigInteger.valueOf(-1);
```
Adjust access modifiers accordingly.
Must these be package private?
Why are the members not declared `final`?

jankoerner replied:
[c7e20f6](https://github.com/ultimate-pa/ultimate/pull/756/commits/c7e20f629801d034fac251784ff79e89a52e26ac)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryArea.java:
```diff
@@ -0,0 +1,12 @@
+package de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.base.chandler;
+
+public enum MemoryArea {
+	STACK, HEAP,;
```
Remove the extra comma to avoid further (unnamed) enum constant.

```suggestion
	STACK, HEAP;
```

jankoerner replied:
[4c4ef60](https://github.com/ultimate-pa/ultimate/pull/756/commits/4c4ef60303828185ffc4cc3fd1cfce1ae892b79b)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryModel.java:
```diff
@@ -0,0 +1,516 @@
+		default:
+			throw new UnsupportedOperationException(memoryStructurePreference + " is an invalid memory structure.");
+		}
+	}
```
This should be outsourced to a memory model factory (following a singleton or builder pattern) like the already existing `MemoryModelFactory` class.

jankoerner replied:
See [c18c10e](https://github.com/ultimate-pa/ultimate/pull/756/commits/c18c10e208a8f1b234ef5a99b3c199c45db9a87f)
[b55fe32](https://github.com/ultimate-pa/ultimate/pull/756/commits/b55fe32ff19db26198aef8b7861493e844f47238)
[b4c15b5](https://github.com/ultimate-pa/ultimate/pull/756/commits/b4c15b57755205bde54a5569ff4346e5fa325133)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryModelDeclarations.java:
```diff
@@ -73,6 +74,12 @@ public enum MemoryModelDeclarations {
+
+	ULTIMATE_STACK_ALLOCATIONS("#StackAllocations"),
+
+	ULTIMATE_HEAP_ALLOCATIONS("#HeapAllocations")
```
The constants like `#InitialAllocations` should be declared in the specific `SFO` class.

jankoerner replied:
[8d3d5d9](https://github.com/ultimate-pa/ultimate/pull/756/commits/8d3d5d978bca688527c4303d05ae838e7ffafd14)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryPointerBase.java:
```diff
@@ -0,0 +1,64 @@
+
+public abstract class MemoryPointerBase implements IMemoryPointer {
+	TypeSizes mTypeSizes;
+	BoogieType mBoogieType;
```
Set the appropriate access modifiers!

jankoerner replied:
[c7e20f6](https://github.com/ultimate-pa/ultimate/pull/756/commits/c7e20f629801d034fac251784ff79e89a52e26ac)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryStructureMultiBitprecise.java:
Underscores in Java filenames should be avoided according to the Java camel case style. This also applies to the class name of this source code file.

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryStructureSingleBitprecise.java:
Underscores in Java filenames should be avoided according to the Java camel case style. This also applies to the class name of this source code file.

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryStructureUnbounded.java:
Underscores in Java filenames should be avoided according to the Java camel case style. This also applies to the class name of this source code file.

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/StaticObjectsHandler.java:
```diff
@@ -85,14 +85,11 @@ public class StaticObjectsHandler {
 
-	private final ILogger mLogger;
-
 	public StaticObjectsHandler(final ILogger logger) {
```
Since the `ILogger` member is removed from the class, the unused parameter in the constructor can also be removed.

```suggestion
	public StaticObjectsHandler() {
```

jankoerner replied:
See [17dd19a](https://github.com/ultimate-pa/ultimate/pull/756/commits/17dd19afa57a31fa314c65c59b02ab8c65618db3)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/PointerValue1D.java:
```diff
@@ -0,0 +1,11 @@
+/**
+ * Represents a single value forming a pointer struct in Boogie.
+ */
+// TODO Shouldn't this already be reassembled in BoogiePreprocessor backtranslation?
```
Check and assess comment. Remove it if possible!

```suggestion
```

jankoerner replied:
[896f3bb](https://github.com/ultimate-pa/ultimate/pull/756/commits/896f3bbb39d646deff1870dd9bfaf8e2396f9cbe)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/PointerValue2D.java:
```diff
@@ -0,0 +1,11 @@
+/**
+ * Represents a pair of values forming a pointer struct in Boogie.
+ */
+// TODO Shouldn't this already be reassembled in BoogiePreprocessor backtranslation?
```
Check and assess comment. Remove it if possible!

```suggestion
```

jankoerner replied:
[896f3bb](https://github.com/ultimate-pa/ultimate/pull/756/commits/896f3bbb39d646deff1870dd9bfaf8e2396f9cbe)

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/IMemoryMetadata.java:
```diff
@@ -0,0 +1,23 @@
+	 *
+	 * @return The declarations.
+	 */
+	List<MemoryModelDeclarations> metaDataDeclarations();
```
Add prefix `get` to interface method name.

```suggestion
	List<MemoryModelDeclarations> getMetaDataDeclarations();
```

jankoerner replied:
See [3cb57aa](https://github.com/ultimate-pa/ultimate/pull/756/commits/3cb57aac1b47ed100add1c3c560b8ae9666ca74c)

jankoerner committed:
CACSL2Boogie: Removed underscore from class names

jankoerner committed:
CACSL2Boogie: Renamed functions in interfaces

jankoerner committed:
CACSL2Boogie: Removed unused parameter

jankoerner committed:
CACSL2Boogie: Removed unused class

jankoerner committed:
CACSL2Boogie: Rename factory method

jankoerner committed:
CACSL2Boogie: Added addressing factory

jankoerner committed:
CACSL2Boogie: Added structure factory

jankoerner committed:
CACSL2Boogie: Added model factory

jankoerner committed:
CACSL2Boogie: Removed unused field

jankoerner committed:
CACSL2Boogie: Removed misleading comma

jankoerner committed:
CACSL2Boogie: Added access modifier

jankoerner committed:
CACSL2Boogie: Consistent naming of base classes

jankoerner committed:
CACSL2Boogie: Removed commented code

jankoerner committed:
CACSL2Boogie: Removed suppress warning

jankoerner committed:
CACSL2Boogie: Moved constants to SFO

jankoerner committed:
CACSL2Boogie: Removed old comment

bahnwaerter commented:
The soundness issues that lead to incorrect results on the following benchmarks should be fixed.

- [x] **Benchmark**: [list-simple/dll2n_append_unequal.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/list-simple/dll2n_append_unequal.yml)
  Program: [list-simple/dll2n_append_unequal.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/list-simple/dll2n_append_unequal.i)
  Actual result: `false(unreach-call)`
  Expected result: `true(unreach-call)`

- [x] **Benchmark**: [aws-c-common/aws_array_list_copy_harness_negated.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/aws-c-common/aws_array_list_copy_harness_negated.yml)
  Program: [aws-c-common/aws_array_list_copy_harness_negated.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/aws-c-common/aws_array_list_copy_harness_negated.i)
  Actual result: `true(unreach-call)`
  Expected result: `false(unreach-call)`

- [x] **Benchmark**: [aws-c-common/aws_string_destroy_secure_harness_negated.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/aws-c-common/aws_string_destroy_secure_harness_negated.yml)
  Program: [aws-c-common/aws_string_destroy_secure_harness_negated.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/aws-c-common/aws_string_destroy_secure_harness_negated.i)
  Actual result: `true(unreach-call)`
  Expected result: `false(unreach-call)`

- [ ] **Benchmark**: [ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--ethernet--sfc--sfc.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--ethernet--sfc--sfc.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.yml)
  Program: [ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--ethernet--sfc--sfc.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--ethernet--sfc--sfc.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.i)
  Actual result: `false(unreach-call)`
  Expected result: `true(unreach-call)`

- [ ] **Benchmark**: [ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--ethernet--sfc--sfc.ko-ldv_main2_sequence_infinite_withcheck_stateful.cil.out.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--ethernet--sfc--sfc.ko-ldv_main2_sequence_infinite_withcheck_stateful.cil.out.yml)
  Program: [ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--ethernet--sfc--sfc.ko-ldv_main2_sequence_infinite_withcheck_stateful.cil.out.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--ethernet--sfc--sfc.ko-ldv_main2_sequence_infinite_withcheck_stateful.cil.out.i)
  Actual result: `false(unreach-call)`
  Expected result: `true(unreach-call)`

- [ ] **Benchmark**: [ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--ethernet--sfc--sfc.ko-ldv_main3_sequence_infinite_withcheck_stateful.cil.out.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--ethernet--sfc--sfc.ko-ldv_main3_sequence_infinite_withcheck_stateful.cil.out.yml)
  Program: [ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--ethernet--sfc--sfc.ko-ldv_main3_sequence_infinite_withcheck_stateful.cil.out.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--ethernet--sfc--sfc.ko-ldv_main3_sequence_infinite_withcheck_stateful.cil.out.i)
  Actual result: `false(unreach-call)`
  Expected result: `true(unreach-call)`

- [ ] **Benchmark**: [ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-sound--core--snd-rawmidi.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-sound--core--snd-rawmidi.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.yml)
  Program: [ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-sound--core--snd-rawmidi.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-sound--core--snd-rawmidi.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.i)
  Actual result: `false(unreach-call)`
  Expected result: `true(unreach-call)`

- [ ] **Benchmark**: [ldv-validator-v0.8/linux-stable-4ee267b-1-130_7a-drivers--hwmon--ibmaem.ko-entry_point_ldv-val-v0.8.cil.out.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-validator-v0.8/linux-stable-4ee267b-1-130_7a-drivers--hwmon--ibmaem.ko-entry_point_ldv-val-v0.8.cil.out.yml)
  Program: [ldv-validator-v0.8/linux-stable-4ee267b-1-130_7a-drivers--hwmon--ibmaem.ko-entry_point_ldv-val-v0.8.cil.out.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-validator-v0.8/linux-stable-4ee267b-1-130_7a-drivers--hwmon--ibmaem.ko-entry_point_ldv-val-v0.8.cil.out.i)
  Actual result: `true(unreach-call)`
  Expected result: `false(unreach-call)`

- [ ] **Benchmark**: [ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--media--pci--pt3--earth-pt3.ko-entry_point.cil.out.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--media--pci--pt3--earth-pt3.ko-entry_point.cil.out.yml)
  Program: [ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--media--pci--pt3--earth-pt3.ko-entry_point.cil.out.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--media--pci--pt3--earth-pt3.ko-entry_point.cil.out.i)
  Actual result: `true(unreach-call)`
  Expected result: `false(unreach-call)`

- [ ] **Benchmark**: [ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--net--xen-netback--xen-netback.ko-entry_point.cil.out.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--net--xen-netback--xen-netback.ko-entry_point.cil.out.yml)
  Program: [ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--net--xen-netback--xen-netback.ko-entry_point.cil.out.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--net--xen-netback--xen-netback.ko-entry_point.cil.out.i)
  Actual result: `true(unreach-call)`
  Expected result: `false(unreach-call)`

- [ ] **Benchmark**: [ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--video--fbdev--core--fb.ko-entry_point.cil.out.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--video--fbdev--core--fb.ko-entry_point.cil.out.yml)
  Program: [ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--video--fbdev--core--fb.ko-entry_point.cil.out.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-linux-4.2-rc1/linux-4.2-rc1.tar.xz-08_1a-drivers--video--fbdev--core--fb.ko-entry_point.cil.out.i)
  Actual result: `false(unreach-call)`
  Expected result: `true(unreach-call)`

- [ ] **Benchmark**: [ldv-linux-3.14/linux-3.14_linux-kernel-locking-mutex_drivers-net-ethernet-broadcom-bnx2.cil.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-linux-3.14/linux-3.14_linux-kernel-locking-mutex_drivers-net-ethernet-broadcom-bnx2.cil.yml)
  Program: [ldv-linux-3.14/linux-3.14_linux-kernel-locking-mutex_drivers-net-ethernet-broadcom-bnx2.cil.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-linux-3.14/linux-3.14_linux-kernel-locking-mutex_drivers-net-ethernet-broadcom-bnx2.cil.i)
  Actual result: `false(unreach-call)`
  Expected result: `true(unreach-call)`

- [ ] **Benchmark**: [ldv-linux-3.14-races/linux-3.14--drivers--usb--misc--adutux.ko.cil.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-linux-3.14-races/linux-3.14--drivers--usb--misc--adutux.ko.cil.yml)
  Program: [ldv-linux-3.14-races/linux-3.14--drivers--usb--misc--adutux.ko.cil.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-linux-3.14-races/linux-3.14--drivers--usb--misc--adutux.ko.cil.i)
  Actual result: `true(unreach-call)`
  Expected result: `false(unreach-call)`

- [ ] **Benchmark**: [ldv-linux-4.0-rc1-mav/linux-4.0-rc1---drivers--char--ipmi--ipmi_msghandler.ko.cil.yml](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-linux-4.0-rc1-mav/linux-4.0-rc1---drivers--char--ipmi--ipmi_msghandler.ko.cil.yml)
  Program: [ldv-linux-4.0-rc1-mav/linux-4.0-rc1---drivers--char--ipmi--ipmi_msghandler.ko.cil.i](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/tree/main/c/ldv-linux-4.0-rc1-mav/linux-4.0-rc1---drivers--char--ipmi--ipmi_msghandler.ko.cil.i)
  Actual result: `false(unreach-call)`
  Expected result: `true(unreach-call)`

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/ArrayHandler.java:
```diff
@@ -281,8 +280,4 @@ private void addArrayBoundsCheckForCurrentIndex(final ILocation loc, final RValu
 
 	}
-
-	private static boolean isInnermostSubscriptExpression(final IASTArraySubscriptExpression node) {
```
Why was this method removed?

jankoerner replied:
I fixed some warnings and this method was unused.
I reverted this change such that it can be removed in another PR.

jankoerner replied:
[eef09c2](https://github.com/ultimate-pa/ultimate/pull/756/commits/eef09c25eaba75b29741176d78ab1c0655e413f6)

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/CCharacterConstant.java:
```diff
@@ -44,6 +44,7 @@ public class CCharacterConstant {
 	/**
 	 * Representation of this character literal in source code.
 	 */
+	@SuppressWarnings("unused")
```
Why did you add these `@SuppressWarnings` (also in other files)? You did not actually change this class, right? If this is unused indeed, it should be rather removed than suppressed, but also not in this PR.

jankoerner replied:
[78914f3](https://github.com/ultimate-pa/ultimate/pull/756/commits/78914f3cc07b8e4527acb7d9e8485e484717cc2c)

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/IMemoryManagementStrategy.java:
```diff
@@ -0,0 +1,54 @@
+package de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.base.chandler;
```
Please add header and author.

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/IMemoryStructure.java:
```diff
@@ -0,0 +1,41 @@
+package de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.base.chandler;
```
Please add header and author.

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryAddressing1D.java:
```diff
@@ -0,0 +1,221 @@
+package de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.base.chandler;
```
Please add header and author.

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryAddressing1D.java:
```diff
@@ -0,0 +1,221 @@
+import de.uni_freiburg.informatik.ultimate.core.model.models.ILocation;
+
+/**
+ * The one dimensional memory addressing.
```
Add some more description

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryAddressing1D.java:
```diff
@@ -0,0 +1,221 @@
+		super(typeHandler, exprTranslation, booleanArrayHelper, typeSizes, typeSizeAndOffsetComputer, pointer);
+
+		final var pointerIntegerMode = settings.getPointerIntegerCastMode();
+		mPointerIntegerConversion = switch (pointerIntegerMode) {
```
As you have a base class for the addressing model: Can you move this code there rather than duplicating it?

jankoerner replied:
[cb24900](https://github.com/ultimate-pa/ultimate/pull/756/commits/cb24900a45a06fade8ac89240af5672186fb11e6)

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryMetadataDefault2D.java:
```diff
@@ -0,0 +1,129 @@
+	 *
+	 * @return The expression.
+	 */
+	@SuppressWarnings("static-method")
```
Again: Don't suppress any warnings (usually), so we can get the warning in Sonar or Eclipse and potentially fix them.

jankoerner replied:
[78914f3](https://github.com/ultimate-pa/ultimate/pull/756/commits/78914f3cc07b8e4527acb7d9e8485e484717cc2c)

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/MemoryPointerBase.java:
```diff
@@ -0,0 +1,64 @@
+	 *
+	 * @return The expression.
+	 */
+	@SuppressWarnings("static-method")
```
Don't suppress warnings

jankoerner replied:
[78914f3](https://github.com/ultimate-pa/ultimate/pull/756/commits/78914f3cc07b8e4527acb7d9e8485e484717cc2c)

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/DataRaceChecker.java:
```diff
@@ -124,6 +124,7 @@ public DataRaceChecker(final AuxVarInfoBuilder auxVarInfoBuilder, final MemoryHa
 	 * @param lrVal
 	 *            The value being read
 	 */
+	@SuppressWarnings({ "unused" })
```
Don't suppress warnings

jankoerner replied:
[78914f3](https://github.com/ultimate-pa/ultimate/pull/756/commits/78914f3cc07b8e4527acb7d9e8485e484717cc2c)

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/DataRaceChecker.java:
```diff
@@ -236,6 +238,7 @@ private static ImmutableList<String> getAccessPath(final LeftHandSide lhs) {
 		}
 	}
 
+	@SuppressWarnings("deprecation")
```
I am not sure, if this method is indeed deprecated

jankoerner replied:
[78914f3](https://github.com/ultimate-pa/ultimate/pull/756/commits/78914f3cc07b8e4527acb7d9e8485e484717cc2c)

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/util/ISOIEC9899TC3.java:
```diff
@@ -61,6 +61,7 @@ public final class ISOIEC9899TC3 {
 	/**
 	 * Message: "Ignored suffix".
 	 */
+	@SuppressWarnings("unused")
```
Don't suppress warnings

jankoerner replied:
[78914f3](https://github.com/ultimate-pa/ultimate/pull/756/commits/78914f3cc07b8e4527acb7d9e8485e484717cc2c)

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/util/SFO.java:
```diff
@@ -142,15 +142,15 @@ public final class SFO {
 	/**
 	 * Sizeof constant prefix "#sizeof~".
 	 */
-	public static final String SIZEOF = "#sizeof~";
```
Why did you rename these constants?

jankoerner replied:
To match the naming of the PRE_OFFSET

schuessf replied:
But you also renamed this 😄

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/LocationFactory.java:
```diff
@@ -122,6 +122,7 @@ public CACSLProjectLocation(final Set<IASTTranslationUnit> set) {
 					Collections.unmodifiableSet(set.stream().map(a -> a.getFilePath()).collect(Collectors.toSet()));
 		}
 
+		@SuppressWarnings("unused")
```
Don't suppress warnings

jankoerner replied:
[78914f3](https://github.com/ultimate-pa/ultimate/pull/756/commits/78914f3cc07b8e4527acb7d9e8485e484717cc2c)

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/interfaces/handler/ITypeHandler.java:
```diff
@@ -188,4 +189,6 @@ public interface ITypeHandler {
 
 	void addLibraryTypes(Map<String, ICType> libraryTypes);
+
+	IMemoryPointer memoryPointer();
```
Please rename this method to start with `get`

jankoerner replied:
[d3e60d5](https://github.com/ultimate-pa/ultimate/pull/756/commits/d3e60d541bc632c8e25cfa1a569ca19272aaa617)

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/chandler/IMemoryPointer.java:
```diff
@@ -0,0 +1,76 @@
+import de.uni_freiburg.informatik.ultimate.core.model.models.ILocation;
+import de.uni_freiburg.informatik.ultimate.plugins.generator.cacsl2boogietranslator.preferences.CACSLPreferenceInitializer.CheckMode;
+
+public interface IMemoryPointer {
```
The name of this interface seems a bit strange to me. Is this really a good name to describe its use?

bahnwaerter replied:
Any suggestions for a better name, @schuessf?

bahnwaerter replied:
I documented the interface and added some explanation to it in order to provide a clearer description of the intended use for this interface.

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/IBacktranslationPointer.java:
```diff
@@ -0,0 +1,14 @@
+package de.uni_freiburg.informatik.ultimate.plugins.generator.cacsl2boogietranslator;
```
Why did you move this (very simple) interface out of CACSL2BoogieBacktranslator?

jankoerner replied:
I moved them out of the CACSL2BoogieBacktranslator because otherwise there would be a cyclic dependency since BacktranslationPointer1D and BacktranslationPointer2D and  would require the CACSL2BoogieBacktranslator which requires either BacktranslationPointer1D or BacktranslationPointer2D.

The same reason for other interfaces as well.

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/IExpressionOrPointer.java:
```diff
@@ -0,0 +1,5 @@
+package de.uni_freiburg.informatik.ultimate.plugins.generator.cacsl2boogietranslator;
```
Why did you move this (very simple) interface out of CACSL2BoogieBacktranslator?

jankoerner replied:
see above

schuessf commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/WrappedExpression.java:
```diff
@@ -0,0 +1,7 @@
+package de.uni_freiburg.informatik.ultimate.plugins.generator.cacsl2boogietranslator;
```
Why did you move this (very simple) interface out of CACSL2BoogieBacktranslator?

jankoerner replied:
see above

schuessf reviewed (commented):
Thank you for your work! I briefly looked over the structure of your new code and it looks good to me -- if it works, we can merge it in principal.

I left a few more general comments. In addition: Please make sure for all the files you added that they the have at least a header, an author and a description such that we know what the class is supposed to do.

jankoerner committed:
CACSL2Boogie: Renamed method

jankoerner committed:
CACSL2Boogie: Fixed some warnings

Removed introduced Supresswarning to allow external tools to notice
them.

jankoerner committed:
CACSL2Boogie: Moved duplicated code to base

Also renamed some classes to match the naming scheme.

jankoerner committed:
CACSL2Boogie: Restored deleted method

This method was deleted because it is unused but restored again such
that it can be deleted in a dedicated commit / pull-request.

schuessf commented:
I just checked the latest nightly and a lot of tests failed, but (probably) only because the new setting "memory neutrality" did not parse an empty string to an empty set and therefore the check in `MemoryPointer1D` always failed. Therefore, I fixed the parsing on `dev` (23b7c4e) and hopefully the test will all succeed in the next nightly.

schuessf commented:
@jankoerner The nightly tests run now in principle. There is however one issue left, where the existing implementation still only works with 2D pointers, see [here](https://github.com/ultimate-pa/ultimate/blob/dev/trunk/source/BoogiePreprocessor/src/de/uni_freiburg/informatik/ultimate/boogie/preprocessor/memoryslicer/AliasAnalysis.java#L442)

Edit: I just fixed that myself.

schuessf committed:
Attempt to fix 1D model in memory slicer

schuessf committed:
Use always 1D model in WitnessRegressionTestSuite, remove duplicate settings/toolchains

schuessf committed:
Remove incompatible setting

schuessf committed:
Use 1D memory model in all SV-COMP settings (except memsafety)

schuessf committed:
CC fixes

schuessf committed:
Add headers

schuessf committed:
Backtranslation: Inline simple declarations in files, add headers

schuessf committed:
Minor: Reduce visibility

schuessf committed:
Revert renaming of fields in SFO, remove unused field

schuessf committed:
Revert adding old code

bahnwaerter committed:
Add Java documentation to new memory model classes

bahnwaerter committed:
Add Java documentation to new memory model interfaces and add author name

schuessf committed:
Make fields private, remove unused fields

schuessf committed:
Move field and methods to base class
